stablehlo/reference/NumPy.cpp

```diff
@@ -0,0 +1,349 @@
+  constexpr char kDescr[] = "'descr':";
+  constexpr int kDescrSize = 8;
+
+  const std::size_t descrOffset = header.find(kDescr);
+  if (descrOffset == std::string::npos) return llvm::errc::invalid_argument;
+
+  const std::size_t needleSize = header.find(',', descrOffset + kDescrSize + 1);
+  std::string typeString = header.substr(
+      descrOffset + kDescrSize, needleSize - (descrOffset + kDescrSize));
+
+  if (typeString.front() != '\'' || typeString.back() != '\'')
+    return llvm::errc::invalid_argument;
+
+  typeString = typeString.substr(1, typeString.size() - 2);
+
+  if (typeString.size() < 3) return llvm::errc::invalid_argument;
+
+  if (getNumPyType<T>() != typeString[1]) return llvm::errc::invalid_argument;
+
+  return std::stoi(typeString.substr(2));
+}
+
+// Parses the `shape` key of the NumPy file format dictionary. Returns a vector
+// representing the parsed shape or errors otherwise.
+static llvm::ErrorOr<ArrayRef<int64_t>> parseShapeHeader(
+    const std::string& header) {
+  const std::size_t shapeOffset = header.find("'shape':");
+  const std::size_t dimEnd = header.find(')', shapeOffset);
+  if (shapeOffset == std::string::npos || dimEnd == std::string::npos)
+    return llvm::errc::invalid_argument;
+
+  // By convention, NumPy writers should include the shape key last (preserving
+  // alphabetical ordering relative to other header keys), however we cannot
+  // assume this to be unilaterally true. Therefore, apply a tight bound on the
+  // regex matching for dimension integrals.
+  std::regex dimRegex("[0-9]+");
+  std::smatch dimMatch;
+  std::string shapeString = header.substr(shapeOffset, shapeOffset - dimEnd);
+  std::vector<int64_t> shape(4);
+
+  while (std::regex_search(shapeString, dimMatch, dimRegex)) {
+    shape.push_back(std::stoi(dimMatch[0]));
+    shapeString = dimMatch.suffix();
+  }
+
+  return shape;
+}
+
+template <typename T>
+static llvm::Error readNumpyHeader(std::ifstream& in, size_t& wordSize,
+                                   ArrayRef<int64_t> shape) {
+  char magicString[kMagicStringSize];
+  if (!in.read(magicString, kMagicStringSize))
+    return llvm::createStringError(llvm::errc::io_error,
+                                   "Failed to read NumPy magic string.");
+
+  if (strncmp(magicString, kMagicString, kMagicStringSize))
+    return llvm::createStringError(llvm::errc::invalid_argument,
+                                   "Invalid NumPy file format detected.");
+
+  // This utility is presently only used from the context of the check dialect
+  // for validation. It suffices to only support NumPy v1.0 for now.
+  char majorVersion, minorVersion;
+  in.read(&majorVersion, 1);
+  in.read(&minorVersion, 1);
+
+  if (majorVersion != kMajorVersion || minorVersion != kMinorVersion)
+    return llvm::createStringError(
+        llvm::errc::invalid_argument,
+        "Invalid NumPy version detected. Only v1.0 supported.");
```

Comment:
Can you provide dynamic error message, something like:
`Invalid NumPy version: %s. Expected version to be %s or %s`